### PR TITLE
feat: decentralized search

### PR DESCRIPTION
**Pull request by hanszhoou**
## Description
https://mask.atlassian.net/browse/MF-2542

Closes # (NO_ISSUE)

## Type of change

- [ ] Documentation
- [ ] Code refactoring (Restructuring existing code w/o changing its observable behavior)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (a fix or feature that would make something no longer possible to do/require old user must upgrade their Mask Network to this new version)

## Checklist

- [ ] My code follows the style guidelines of this project.
- [ ] I have performed a self-review of my own code.
  - [ ] I have removed all in development `console.log`s
  - [ ] I have removed all commented code.
- [ ] I have commented on my code, particularly in hard-to-understand areas.
- [ ] I have read [Internationalization Guide](https://github.com/DimensionDev/Maskbook/blob/develop/docs/i18n-guide.md) and moved text fields to the i18n JSON file.

If this PR depends on external APIs:

- [ ] I have configured those APIs with CORS headers to let extension requests get passed. 
  - chrome extension: `chrome-extension://[id]`
  - firefox extension: `moz-extension://[id]`
- [ ] I have delegated all web requests to the background service via the internal RPC bridge.

### CODE REVIEW

File: packages/plugins/DecentralizedSearch/src/SNSAdaptor/SearchResultInspector.tsx [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+    content: {
+        position: 'relative',
+        background: theme.palette.common.white,
+        maxHeight: 478,
+        overflow: 'scroll',
+    },
+}))
+
+export function SearchResultInspector(props: { keyword: string }) {
+    const ENS_Plugin = useActivatedPluginSNSAdaptor(PluginID.ENS, 'any')
```
Comment (guanbinrui) [inline packages/plugins/DecentralizedSearch/src/SNSAdaptor/SearchResultInspector.tsx:38]:
```suggestion
    const ENS_Plugin = useActivatedPluginSNSAdaptor(PluginID.DecentralizedSearch, 'any')
```

File: packages/plugins/DecentralizedSearch/src/SNSAdaptor/SearchResultInspector.tsx [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+    tabs: {
+        display: 'flex',
+        position: 'relative',
+        padding: theme.spacing(0, 2),
+    },
+    content: {
+        position: 'relative',
+        background: theme.palette.common.white,
+        maxHeight: 478,
+        overflow: 'scroll',
```
Comment (guanbinrui) [inline packages/plugins/DecentralizedSearch/src/SNSAdaptor/SearchResultInspector.tsx]:
```suggestion
        overflow: 'auto',
```

File: packages/mask/src/plugins/Trader/SNSAdaptor/index.tsx [hunk 3]
```diff
@@ -37,7 +37,7 @@ const sns: Plugin.SNSAdaptor.Definition<
         },
         Utils: {
             shouldDisplay(keyword: string) {
-                return /[#$]\w+/.test(keyword)
+                return /[#$]\w+/.test(keyword) || (isValidAddress(keyword) && !isZeroAddress(keyword))
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Trader/SNSAdaptor/index.tsx:40]: Do we need to support `ENS`?

File: packages/web3-shared/base/src/specs/index.ts [hunk 4]
```diff
@@ -1533,3 +1533,8 @@ export interface Web3UI<ChainId, ProviderType, NetworkType> {
         ProviderIconClickBait?: React.ComponentType<ProviderIconClickBaitProps<ChainId, ProviderType, NetworkType>>
     }
 }
+
+export enum SearchKeywordType {
+    NameServiceDomain = 'NameServiceDomain',
```
Comment (guanbinrui) [inline packages/web3-shared/base/src/specs/index.ts]:
```suggestion
    Domain = 'Domain',
```

File: packages/mask/src/plugins/Trader/SNSAdaptor/index.tsx (hunk 3 above)
Comment (hanszhoou) [reply packages/mask/src/plugins/Trader/SNSAdaptor/index.tsx:40]: At Trader side, no.

File: packages/plugins/DecentralizedSearch/src/SNSAdaptor/SearchResultInspector.tsx (hunk 1 above)
Comment (guanbinrui) [inline packages/plugins/DecentralizedSearch/src/SNSAdaptor/SearchResultInspector.tsx:37]: Don't forget to add `DecentralizedSearch` to `PluginID`.

File: packages/plugins/DecentralizedSearch/src/SNSAdaptor/SearchResultInspector.tsx [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+        setLoading(!ensRef.current || ensRef.current?.isLoading)
+        setHiddenAll(Boolean(ensRef.current && ensRef.current?.reversedAddress === undefined) || isLoadingAddressType)
+        setEmpty(
+            Boolean(
+                ensRef.current &&
+                    (!ensRef.current?.reversedAddress || !ensRef.current?.tokenId) &&
+                    !ensRef.current?.isLoading,
+            ),
+        )
+        setError(Boolean(ensRef.current?.isError))
```
Comment (guanbinrui) [inline packages/plugins/DecentralizedSearch/src/SNSAdaptor/SearchResultInspector.tsx]: Avoid using `setState` in a `useEffect`. They can be computed.

File: packages/plugins/DecentralizedSearch/src/SNSAdaptor/SearchResultInspector.tsx [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+        isLoadingAddressType ||
+        (!addressType && ensRef.current?.reversedAddress && !Others?.isValidDomain?.(props.keyword))
+    )
+        return null
+
+    return (
+        <Hidden hidden={isHiddenAll}>
+            <DecentralizedSearchPostExtraInfoWrapper>
+                <Hidden hidden={isLoading || isEmpty || isError}>
+                    <ENS_SearchResult
```
Comment (guanbinrui) [inline packages/plugins/DecentralizedSearch/src/SNSAdaptor/SearchResultInspector.tsx:107]: Better avoid dead-coded to ENS result. Any plugin can define itself as a search result.

File: packages/web3-hooks/evm/src/useFungibleTokenBaseOnChainIdList.ts [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+import { attemptUntil } from '@masknet/web3-shared-base'
+import type { ChainId } from '@masknet/web3-shared-evm'
+
+export function useFungibleTokenBaseOnChainIdList<
+    S extends 'all' | void = void,
+    T extends NetworkPluginID = NetworkPluginID,
+>(
+    pluginID?: T,
+    address?: string,
+    chainIdList?: ChainId[],
```
Comment (guanbinrui) [inline packages/web3-hooks/evm/src/useFungibleTokenBaseOnChainIdList.ts]:
```suggestion
    chainIdList?: Array<Web3Helper.ChainId[T]>,
```

File: packages/web3-hooks/evm/src/useFungibleTokenBaseOnChainIdList.ts [hunk 8]
```diff
@@ -0,0 +1,35 @@
+import { useAsyncRetry } from 'react-use'
+import { flatten } from 'lodash-es'
+import type { NetworkPluginID } from '@masknet/shared-base'
+import type { Web3Helper } from '@masknet/web3-helpers'
+import { useWeb3Connection, useWeb3State, useChainId } from '@masknet/web3-hooks-base'
+import { attemptUntil } from '@masknet/web3-shared-base'
+import type { ChainId } from '@masknet/web3-shared-evm'
+
+export function useFungibleTokenBaseOnChainIdList<
```
Comment (guanbinrui) [inline packages/web3-hooks/evm/src/useFungibleTokenBaseOnChainIdList.ts]: Move to `@masknet/web3-shared-base`.

File: packages/web3-hooks/evm/src/useAddressTypeBaseOnChainIdList.ts [hunk 9]
```diff
@@ -0,0 +1,22 @@
+import { useAsyncRetry } from 'react-use'
+import type { NetworkPluginID } from '@masknet/shared-base'
+import { attemptUntil } from '@masknet/web3-shared-base'
+import type { Web3Helper } from '@masknet/web3-helpers'
+import { useWeb3Connection, useWeb3State } from '@masknet/web3-hooks-base'
+import type { ChainId } from '@masknet/web3-shared-evm'
+
+export function useAddressTypeBaseOnChainIdList<
```
Comment (guanbinrui) [inline packages/web3-hooks/evm/src/useAddressTypeBaseOnChainIdList.ts]: Move to `@masknet/web3-shared-base`.

File: packages/web3-hooks/evm/src/useAddressTypeBaseOnChainIdList.ts [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+    S extends 'all' | void = void,
+    T extends NetworkPluginID = NetworkPluginID,
+>(pluginID?: T, address?: string, chainIdList?: ChainId[], options?: Web3Helper.Web3ConnectionOptionsScope<S, T>) {
```
Comment (guanbinrui) [inline packages/web3-hooks/evm/src/useAddressTypeBaseOnChainIdList.ts]:
```suggestion
>(pluginID?: T, address?: string, chainIdList?: Array<Web3Helper.ChainId[T]>, options?: Web3Helper.Web3ConnectionOptionsScope<S, T>) {
```

File: packages/plugins/DecentralizedSearch/src/SNSAdaptor/SearchResultInspector.tsx (hunk 1 above)
Comment (hanszhoou) [reply packages/plugins/DecentralizedSearch/src/SNSAdaptor/SearchResultInspector.tsx:37]: Here we use ENS plugin in the DecentralizedSearch.

File: packages/plugins/DecentralizedSearch/src/SNSAdaptor/SearchResultInspector.tsx (hunk 6 above)
Comment (hanszhoou) [reply packages/plugins/DecentralizedSearch/src/SNSAdaptor/SearchResultInspector.tsx:107]: It is defined by ENS plugin.

**Review by guanbinrui — APPROVED** [review]

**Pull request merged**